File: linda-core/src/main/java/it/unibo/coordination/utils/events/AsyncEventEmitter.java
package it.unibo.coordination.utils.events;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;

public interface AsyncEventEmitter<Arg> extends SyncEventEmitter<Arg> {

    static <A> AsyncEventEmitter<A> ordered(ExecutorService executor) {
        return new AsyncOrderedEventSourceImpl<>(executor);
    }

    CompletableFuture<Arg> asyncEmit(Arg event);
}

File: utils/src/main/kotlin/it/unibo/coordination/utils/events/AsyncEventEmitter.kt
package it.unibo.coordination.utils.events

import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutorService

interface AsyncEventEmitter<T> : EventEmitter<T> {

    fun asyncEmit(event: T): CompletableFuture<T>

    @JvmDefault
    override fun emit(event: T) {
        asyncEmit(event)
    }

    companion object {

        @JvmStatic
        fun <A> ordered(executor: ExecutorService): AsyncEventEmitter<A> {
            return AsyncOrderedEventSource(executor)
        }
    }
}